File: services/api/app/api/routes/notifications.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import AsyncGenerator

from app.api.deps import get_current_user
from app.api.rate_limit import rate_limiter
from app.core.config import settings
from app.core.redis import get_redis_async
from app.crud.notifications import list_notifications, mark_all_read, mark_read, unread_count
from app.db.session import get_db
from app.schemas.notifications import (
    NotificationListOut,
    NotificationOut,
    PageOut,
    UnreadCountOut,
)
from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

router = APIRouter(prefix="/v1", tags=["notifications"])


@router.get(
    "/notifications",
    response_model=NotificationListOut,
    dependencies=[Depends(rate_limiter("notifications", limit=120, window_seconds=60))],
)
def get_notifications(
    *,
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    unread_only: bool = Query(False),
    db: Session = Depends(get_db),
    user=Depends(get_current_user),
):
    total, rows = list_notifications(
        db,
        user_id=user.id,
        unread_only=unread_only,
        limit=limit,
        offset=offset,
    )

    items = [
        NotificationOut(
            id=r.event.id,
            created_at=r.event.created_at,
            read_at=r.event.read_at,
            shelf_item_id=r.event.shelf_item_id,
            title=r.title,
            author=r.author,
            format=r.event.format,
            old_status=r.event.old_status,
            new_status=r.event.new_status,
            deep_link=r.event.deep_link,
        )
        for r in rows
    ]

    return NotificationListOut(page=PageOut(limit=limit, offset=offset, total=total), items=items)


@router.get(
    "/notifications/unread-count",
    response_model=UnreadCountOut,
    dependencies=[Depends(rate_limiter("notifications_unread", limit=240, window_seconds=60))],
)
def get_unread_count(
    db: Session = Depends(get_db),
    user=Depends(get_current_user),
):
    return UnreadCountOut(unread=unread_count(db, user_id=user.id))


@router.post(
    "/notifications/{notification_id}/read",
    status_code=204,
    dependencies=[Depends(rate_limiter("notifications_mark_read", limit=240, window_seconds=60))],
)
def mark_notification_read(
    notification_id: str,
    db: Session = Depends(get_db),
    user=Depends(get_current_user),
):
    ok = mark_read(db, user_id=user.id, notification_id=notification_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Notification not found")
    return None


@router.post(
    "/notifications/mark-all-read",
    dependencies=[Depends(rate_limiter("notifications_mark_all", limit=60, window_seconds=60))],
)
def mark_all_notifications_read(
    db: Session = Depends(get_db),
    user=Depends(get_current_user),
):
    n = mark_all_read(db, user_id=user.id)
    return {"updated": n}


@router.get("/notifications/events")
async def stream_notifications(
    user=Depends(get_current_user),
) -> StreamingResponse:
    redis_client = get_redis_async(settings.redis_url)
    channel = f"notify:{user.id}"

    async def event_generator() -> AsyncGenerator[str, None]:
        pubsub = redis_client.pubsub()
        await pubsub.subscribe(channel)

        # Basic keepalive to help proxies (optional)
        last_keepalive = datetime.now(timezone.utc)

        try:
            async for message in pubsub.listen():
                now = datetime.now(timezone.utc)

                # Keepalive comment every ~25s
                if (now - last_keepalive).total_seconds() > 25:
                    yield ": keepalive\n\n"
                    last_keepalive = now

                if message is None:
                    continue
                if message.get("type") != "message":
                    continue

                data_raw = message.get("data")
                if isinstance(data_raw, (bytes, bytearray)):
                    data_str = data_raw.decode("utf-8")
                else:
                    data_str = str(data_raw)

                payload = json.loads(data_str)
                payload.setdefault("ts", now.isoformat())

                yield f"data: {json.dumps(payload)}\n\n"
        finally:
            await pubsub.unsubscribe(channel)
            await pubsub.close()

    return StreamingResponse(event_generator(), media_type="text/event-stream")

File: shelfsync-api/src/routes/notifications.ts
import type { Hono } from "hono";
import { jsonResponse } from "../utils/http";
import { parseLimitParam, parseOffsetParam, parseBooleanQuery } from "../utils/query";
import { requireAuthUser } from "../auth/session";
import {
	countUnreadNotifications,
	insertNotificationEvent,
	listNotificationsForUser,
	markAllNotificationsRead,
	markNotificationRead,
	resolveShelfItemForNotification,
} from "../db/notifications";
import { publishNotificationEvent } from "../notificationEvents";
import type { Env } from "../types";

export const registerNotificationRoutes = (app: Hono<{ Bindings: Env }>) => {
	app.get("/v1/notifications", async (c) => {
		const user = await requireAuthUser(c);
		if (!user) {
			return jsonResponse({ ok: false, error: "Unauthorized" }, { status: 401 });
		}
		const limit = parseLimitParam(c.req.query("limit"));
		const offset = parseOffsetParam(c.req.query("offset"));
		const unreadOnly = parseBooleanQuery(c.req.query("unread_only"));
		const { total, rows } = await listNotificationsForUser(
			c.env.DB,
			user.id,
			limit,
			offset,
			unreadOnly,
		);
		const items = rows.map((row) => ({
			id: row.id,
			created_at: row.created_at,
			read_at: row.read_at,
			shelf_item_id: row.shelf_item_id,
			title: row.title,
			author: row.author,
			format: row.format,
			old_status: row.old_status,
			new_status: row.new_status,
			deep_link: row.deep_link,
		}));
		return jsonResponse({
			ok: true,
			data: {
				page: { total, limit, offset },
				items,
			},
		});
	});

	app.get("/v1/notifications/unread-count", async (c) => {
		const user = await requireAuthUser(c);
		if (!user) {
			return jsonResponse({ ok: false, error: "Unauthorized" }, { status: 401 });
		}
		const unread = await countUnreadNotifications(c.env.DB, user.id);
		return jsonResponse({ ok: true, data: { unread } });
	});

	app.post("/v1/notifications/:id/read", async (c) => {
		const user = await requireAuthUser(c);
		if (!user) {
			return jsonResponse({ ok: false, error: "Unauthorized" }, { status: 401 });
		}
		const notificationId = c.req.param("id");
		if (!notificationId) {
			return jsonResponse({ ok: false, error: "Notification ID is required" }, { status: 400 });
		}
		const updated = await markNotificationRead(c.env.DB, user.id, notificationId);
		if (!updated) {
			return jsonResponse({ ok: false, error: "Notification not found" }, { status: 404 });
		}
		return new Response(null, { status: 204 });
	});

	app.post("/v1/notifications/mark-all-read", async (c) => {
		const user = await requireAuthUser(c);
		if (!user) {
			return jsonResponse({ ok: false, error: "Unauthorized" }, { status: 401 });
		}
		const updated = await markAllNotificationsRead(c.env.DB, user.id);
		return jsonResponse({ ok: true, data: { updated } });
	});

	app.post("/v1/notifications/events/test", async (c) => {
		const user = await requireAuthUser(c);
		if (!user) {
			return jsonResponse({ ok: false, error: "Unauthorized" }, { status: 401 });
		}
		const body = await c.req.json().catch(() => null);
		const format = typeof body?.format === "string" ? body.format.trim() : "";
		const oldStatus = typeof body?.old_status === "string" ? body.old_status.trim() : "";
		const newStatus = typeof body?.new_status === "string" ? body.new_status.trim() : "";
		if (!format || !oldStatus || !newStatus) {
			return jsonResponse({ ok: false, error: "format, old_status, and new_status are required" }, { status: 400 });
		}
		let shelfItemData;
		try {
			shelfItemData = await resolveShelfItemForNotification(c.env.DB, user.id, {
				shelf_item_id: typeof body?.shelf_item_id === "string" ? body.shelf_item_id : undefined,
				title: typeof body?.title === "string" ? body.title : undefined,
				author: typeof body?.author === "string" ? body.author : undefined,
			});
		} catch (error) {
			const message = error instanceof Error ? error.message : "Shelf item not found";
			return jsonResponse({ ok: false, error: message }, { status: 400 });
		}
		const deepLink = typeof body?.deep_link === "string" ? body.deep_link : null;
		const eventId = await insertNotificationEvent(c.env.DB, {
			userId: user.id,
			shelfItemId: shelfItemData.id,
			format,
			oldStatus,
			newStatus,
			deepLink,
		});
		await publishNotificationEvent(c.env.NOTIFICATION_EVENTS, user.id, {
			id: eventId,
			shelf_item_id: shelfItemData.id,
			title: shelfItemData.title,
			author: shelfItemData.author,
			format,
			old_status: oldStatus,
			new_status: newStatus,
			deep_link: deepLink,
		});
		return jsonResponse({ ok: true, data: { id: eventId } });
	});
};
